File: 733-flood-fill/733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        if image[sr][sc] == color: return image
        rows, cols = len(image), len(image[0])
        dir = [(1, 0), (-1, 0), (0, -1), (0, 1)]
        initial_color = image[sr][sc]
        
        def dfs(u):
            image[u[0]][u[1]] = color
            for di, dj in dir:
                x, y = u[0] + di, u[1] + dj
                if 0 <= x < rows and 0 <= y < cols and image[x][y] == initial_color:
                    dfs((x, y) )
        
        dfs((sr, sc) )
        return image
            

File: 733-flood-fill/733-flood-fill.cpp
#include <vector>

using namespace std;

class Solution {
public:
    void dfs(vector<vector<int> >& image, vector<pair<int, int> > dir, pair<int, int> u, int init_color, int color) {
        image[u.first][u.second] = color;
        for (pair<int, int> v : dir) {
            int x = v.first + u.first, y = v.second + u.second;
            if (0 <= x && x < image.size() && 0 <= y && y < image[0].size() && image[x][y] == init_color)
                dfs(image, dir, make_pair(x, y), init_color, color);
        }
        return;
    }
    
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if (image[sr][sc] == color) return image;
        int rows = image.size(), cols = image[0].size();
        vector<pair<int, int> > dir{make_pair(0, 1), make_pair(0, -1), make_pair(1, 0), make_pair(-1, 0)};
        int initial_color = image[sr][sc];
        dfs(image, dir, make_pair(sr, sc), initial_color, color);
        return image;
    }
};